### PR TITLE
Add configuration cloning

### PR DESCRIPTION
**Pull request by Aaron2000119**
Part of https://github.com/Gamify-IT/issues/issues/387

Adds the option to clone a configuration.

This backend is the same as the towercrush- and finitequiz-backend.

To test use the description in the overworld-backend pull request.

### CODE REVIEW

File: src/main/java/de/unistuttgart/chickenshockbackend/data/Configuration.java [hunk 1]
```diff
@@ -63,4 +63,17 @@ public void addQuestion(final Question question) {
...
     }
+
+    @Override
+    public Object clone() {
+        Configuration config;
+        try {
+            config = (Configuration) super.clone();
+        } catch (CloneNotSupportedException e) {
+            config = new Configuration(this.getQuestions(), this.time);
+        }
```
Comment (delvh) [inline src/main/java/de/unistuttgart/chickenshockbackend/data/Configuration.java]:
```suggestion
        final Configuration config = new Configuration(this.getQuestions(), this.time);
```

File: src/main/java/de/unistuttgart/chickenshockbackend/data/Question.java [hunk 2]
```diff
@@ -56,4 +57,13 @@ public Question(final String text, final String rightAnswer, final Set<String> w
         this.rightAnswer = rightAnswer;
         this.wrongAnswers = wrongAnswers;
     }
+
+    @Override
+    public Object clone() {
+        try {
+            return super.clone();
+        } catch (CloneNotSupportedException e) {
+            return new Question(this.text, this.rightAnswer, new HashSet<>(this.wrongAnswers) {});
```
Comment (delvh) [inline src/main/java/de/unistuttgart/chickenshockbackend/data/Question.java]:
```suggestion
        return new Question(this.text, this.rightAnswer, new HashSet<>(this.wrongAnswers));
```

File: src/test/java/de/unistuttgart/chickenshockbackend/ConfigControllerTest.java [hunk 3]
```diff
@@ -311,4 +311,31 @@ void updateQuestionFromExistingConfigurationNotFound() throws Exception {
...
+        cloneConfig
+            .getQuestions()
+            .forEach(question ->
+                initialConfig
+                    .getQuestions()
+                    .forEach(initialQuestion -> {
+                        assertNotEquals(question.getId(), initialQuestion.getId());
+                    })
+            );
+        assertNotEquals(cloneConfig, initialConfig);
```
Comment (delvh) [inline src/test/java/de/unistuttgart/chickenshockbackend/ConfigControllerTest.java]: Erm… This test looks somewhat irrelevant in this implementation.
You should also check that all questions are still present, and that the db also knows about this new configuration.

Comment (Aaron2000119) [reply src/test/java/de/unistuttgart/chickenshockbackend/ConfigControllerTest.java]: This test, if the cloned questions are a deep copy. I think its important to test especially if the implementation gets changed.

File: src/main/java/de/unistuttgart/chickenshockbackend/data/Configuration.java [hunk 4]
```diff
@@ -63,4 +63,12 @@ public void addQuestion(final Question question) {
     public void removeQuestion(final Question question) {
         this.questions.remove(question);
     }
+
+    @Override
+    public Configuration clone() {
+        return new Configuration(
+            this.questions.stream().map(question -> question = question.clone()).collect(Collectors.toSet()),
```
Comment (delvh) [inline src/main/java/de/unistuttgart/chickenshockbackend/data/Configuration.java]:
```suggestion
            this.questions.stream().map(Question::clone).collect(Collectors.toSet()),
```

File: src/main/java/de/unistuttgart/chickenshockbackend/data/Question.java [hunk 5]
```diff
@@ -56,4 +57,9 @@ public Question(final String text, final String rightAnswer, final Set<String> w
         this.rightAnswer = rightAnswer;
         this.wrongAnswers = wrongAnswers;
     }
+
+    @Override
+    public Question clone() {
+        return new Question(this.text, this.rightAnswer, new HashSet<>(this.wrongAnswers) {});
```
Comment (delvh) [inline src/main/java/de/unistuttgart/chickenshockbackend/data/Question.java]:
```suggestion
        return new Question(this.text, this.rightAnswer, new HashSet<>(this.wrongAnswers));
```

**Review by delvh — APPROVED** [review]
